streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala

```diff
@@ -351,6 +351,50 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
   }
 
   /**
+   * :: Experimental ::
+   * Return a new DStream of data generated by combining the key-value data in `this` stream
+   * with a continuously updated per-key state. The user-provided state tracking function is
+   * applied on each keyed data item along with its corresponding state. The function can choose to
+   * update/remove the state and return a transformed data, which forms the
+   * [[org.apache.spark.streaming.dstream.EmittedRecordsDStream]].
+   *
+   * The specifications of this transformation is made through the
+   * [[org.apache.spark.streaming.StateSpec StateSpec]] class. Besides the tracking function, there
+   * are a number of optional parameters - initial state data, number of partitions, timeouts, etc.
+   * See the [[org.apache.spark.streaming.StateSpec StateSpec spec docs]] for more details.
+   *
+   * Scala example of using `trackStateByKey`:
+   * {{{
+   *    def trackingFunction(key: String, data: Option[Int], wrappedState: State[Int]): Option[Int] = {
+   *      // Check state exists, accordingly update/remove state and return data to emit
+   *    }
+   *
+   *    val spec = StateSpec(trackingFunction).numPartitions(10)
+   *
+   *    val emittedRecordDStream = keyValueDStream.trackStateByKey[StateType, EmittedDataType](spec)
+   *
+   * }}}
+   *
+   * Java example of using `trackStateByKey`:
+   * {{{
+   *      TODO(@zsxwing)
+   * }}}
```

Comment:
The Java example should only be in the Java API, doesn't make sense to have it here